### PR TITLE
Week 4 assessment marcelo

### PR DESCRIPTION
**Pull request by celoazevedo**

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -18,40 +18,120 @@
...
-var colors1 = ["purple", "blue", "green", "yellow", "pink"]
+// var colors1 = ["purple", "blue", "green", "yellow", "pink"]
 // Expected output example (can be a different order): ["yellow", "blue", "pink", "green"]
-var colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saffron"]
+// var colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saffron"]
 // Expected output example (can be a different order): ["saffron", "aquamarine", "periwinkle", "indigo", "ochre"]
 
-
+fdescribe("removeFirstAndShuffle", () => {
+  let colors1 = ["purple", "blue", "green", "yellow", "pink"] // Is declaring the variables inside of the describe method best practice?
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:27]: It's okay to do that

File: code-challenges.test.js [hunk 2]
```diff
@@ -18,40 +18,120 @@
...
+  test("returns an array without the first element and than shuffled", () => {
+      
+    console.log('color1 array -> after test() before expect():' ,colors1)
+      
+    expect(removeFirstAndShuffle(colors1)).toContain("yellow", "blue", "pink", "green") // here we are checking if the result array after running through our removeFirstAndShuffle function contains all of the elements from the original array minus the element at index 0. --> .toContain - Jest method - checks if it contains the data that we pass in. In this case we need to pass in the strings without being inside of an array. I think that this is because istead of checking if the result array has the strings, it checks if the result contains an array with the specific array of strings that we pass in. 
+    expect(removeFirstAndShuffle(colors1)).not.toEqual(["blue", "green", "yellow", "pink"]) // comparing the result with the original  array without the first element And NOT SHUFFLED. --> passed because the arrays should not be the same -- the result is shuffled -->  used the .not.toEaqual
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:37]: These tests together provide solid coverage! good work!

File: code-challenges.test.js [hunk 3]
```diff
@@ -18,40 +18,120 @@
...
+// use .shift() built in array method to remove the first element in the array !!-- Replaced with .slice() and assigned the result to a variable.
+// use .sort(() => Math.random() - 0.5) ---> to shuffle the array
+// return the new shorter and shuffled array
+
+const removeFirstAndShuffle = (arr) => {
+  // arr.shift()
+  let newArr = arr.slice(1)
+  // console.log(newArr)
+  console.log(newArr.sort(() => Math.random() - 0.5))
+  return newArr.sort(() => Math.random() - 0.5)
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:64]: Great use of the sort method and the call back function to set up your randomization

File: code-challenges.test.js [hunk 4]
```diff
@@ -18,40 +18,120 @@
...
 
-
-
 // b) Create the function that makes the test pass.
-
-
+// create a function
+// declare a empty array to push max and min values
+// use the .min and .max from the Math object to get the max and min values from the array
+// !!-- alternative: we could sort the array and grab (push to a new array) the first and last elements of the sorted array. but I think the spread operator is pretty cool... - not sure if there are any pitfalls here but the test is passing.😵‍💫
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:93]: love it!

File: code-challenges.test.js [hunk 5]
```diff
@@ -18,40 +18,120 @@
...
+// return the newArr
+
+const minAndMax = (arr) => {
+  let newArr = []
+  // here we used the spread operator -- it allows the built in methods to check all of the elements in the array -- 'almost like iterating and eveluating each element of the array'
+  newArr.push(Math.min(...arr), Math.max(...arr))
+  return newArr
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:101]: Great work

File: code-challenges.test.js [hunk 6]
```diff
@@ -18,40 +18,120 @@
...
+// declare a variable to store the concatenation of the two arrays --> use .concat()
+// use new ES6 syntax to return an array from the (selection of only unique elements done by using the 'new Set()' es6 syntax) --- the test passed. Please let me know if there are any pitfalls by using this syntax.
+
+// const concatArraysNoDups = (arr1, arr2) => {
+//   let concArr = arr1.concat(arr2)
+//   return Array.from(new Set(concArr))
+// }
+
+// using the spread operator
+const concatArrUniqueVals = (...args) => {
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:133]: ooooo, nice syntax! Smooth like james bond!

File: code_challenges.rb [hunk 7]
```diff
@@ -20,15 +29,59 @@
 album3 = 'Abbey Road'
 # Expected output: 'bby Rd'
 
+def remove_vowels str
+    str.delete "aeiouAEIOU" #delete is a string method that will remove all matching characters!
+end
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:34]: Great!

File: code_challenges.rb [hunk 8]
```diff
@@ -8,7 +8,16 @@
 # Expected output: ['coffee', 'soda water']
 letter_t = 't'
 # Expected output: ['tea', 'water', 'soda water']
+def letter_included (arr, letter)
+    arr.select do |value|
+        value.include?letter
+    end
+end
 
+# refactor with arr.any? { |letter| s.include? letter } # Will return true if any element is present
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:17]: Good idea

File: code_challenges.rb [hunk 9]
```diff
@@ -20,15 +29,59 @@
...
+        if current_speed >= num
+            @current_speed = @current_speed - num
+        else
+            @current_speed = 0
+        end
+    end
+
+end
+
+my_bike = Bike.new 'BMW'
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:71]: Great Work! do you have a bmw bike that's so cool! I want an old school porsche bike

File: interview-questions.md [hunk 10]
```diff
@@ -7,60 +7,60 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 1. What is an instance variable in Ruby? How is it different from other variables?
 
-  Your answer:
+  Your answer:  the instance variable (declared with @) in ruby is declared when you create a class. That will allow every object that is an instance of that class to have access to that variable. It becomes part of the objects.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:10]: Nice!

File: interview-questions.md [hunk 11]
```diff
@@ -7,60 +7,60 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-  Researched answer:
+  Researched answer: Blocks are like anonymous functions in JS. We create blocks by starting with 'do' and ending with 'end'. We can also use {} to create a block. Blocks are a set of instructions (code) that is executed by method. 
 
 
 
 3. Ruby has an implicit return. What does that mean?
 
-  Your answer:
+  Your answer: that we do not need the keyword 'return' to return the result from a block or statement. it is implicit that the function or block will be returning something.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:26]: Yes!

File: interview-questions.md [hunk 12]
```diff
@@ -7,60 +7,60 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-  Researched answer:
+  Researched answer: it means that the last line inside of a method will always be returned (without the need of the return keyword)
 
 
 
 4. What is object-oriented programming? How is it different than functional programming?
 
-  Your answer:
+  Your answer: OOP is a programing paradigm. It is a set of rules that was agreed upon programers to structure the code. The four pillars that support the OOP paradigm are encapsulation, abstraction, inheritance and polymorphism. In functional programming we have the logic wrapped in functions that can be called as needed. That can become very complex / difficult to read / make changes. In oop we define the logic inside of classes and objects and this allows us to easily find the pieace of code that we would like to change, change that code without loosing sleep at night, or add a functionality without having to repeat your code. With functional programing we have our functions and variables 'floating around'.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:34]: This is a great answer to hang on to if you ever get a chance to use it in an interview!

**Review by WalkerTexas-coder — APPROVED** [review]
Marceloooo! Amazing work here. I appreciate all of the different ideas you tried out especially because you documented them so well. Great work all around. Really solid ideas with the testing coverage!

File: code-challenges.test.js (hunk 1 above)
Comment (celoazevedo) [reply code-challenges.test.js:27]: Thank you for responding and for all of the feedback!

File: code-challenges.test.js (hunk 6 above)
Comment (celoazevedo) [reply code-challenges.test.js:133]: Hehehe!

File: code_challenges.rb (hunk 9 above)
Comment (celoazevedo) [reply code_challenges.rb:71]: I don't but I do have a bicycle!

**Pull request merged**